packages/protocol/contracts/psp/AirnodeSponsorPspAllocator.sol

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.6;
+
+import "../rrp/interfaces/IAirnodeRrp.sol";
+import "./AirnodePsp.sol";
+
+// It would be nice if slots could also expire, similar to whitelisting expiration
+contract AirnodeSponsorPspAllocator {
+    struct Slot {
+        bytes32 subscriptionId;
+        address sponsor;
+    }
+
+    IAirnodeRrp public airnodeRrp;
+    AirnodePsp public airnodePsp;
+    mapping(address => mapping(address => uint256))
+        public airnodeToSponsorToNoOfTotalSlots;
+    mapping(address => mapping(address => uint256))
+        public airnodeToSponsorToNoOfUsedSlots;
+    mapping(address => mapping(uint256 => Slot))
+        public airnodeToSlotIndexToSlot;
+
+    constructor(address airnodeRrp_, address airnodePsp_) {
+        airnodeRrp = IAirnodeRrp(airnodeRrp_);
+        airnodePsp = AirnodePsp(airnodePsp_);
+    }
+
+    function setTotalNoOfSlots(
+        address airnode,
+        address sponsor,
+        uint256 noOfTotalSlots
+    ) external {
+        // Check if the caller is an Airnode admin, i.e., this contract will inherit SelfAdminnable
+        // require(msg.sender == ...)
+        airnodeToSponsorToNoOfTotalSlots[airnode][sponsor] = noOfTotalSlots;
+    }
+
+    // Need to come up with a way to prevent people from constantly switching between slots to grief
```

Comment:
You mean like checking if mapping already has a subscriptionId set for airnode/sponsor pair in different slots?

Reply:
I mean say I'm a regular user that wants to subscribe. When I send a transaction that specifies a slot, someone that also has a slot available makes a transaction with a high gas price, frontrunning to get my slot. My tx gets reverted and I see that the slot I attempted to get is now unavailable so I make another tx for another slot. The attacker quickly unsubscribes and subscribes again for the next slot I attempted to get, and this keeps happening.